### ISSUE TITLE
Residual asymmetry while fitting point source

### ISSUE THREAD

**Issue by ke27whal** [pysersic/pysersic#66]
Hello there!

I'm trying to run through a sanity check by fitting a point source with a point source model, and the resulting residuals are asymmetric. How are xc and yc defined? Are they pixel centers? Is this expected?

I'll attach the code to confirm this is reproducible and not just an issue I'm having.

```
import pysersic
import numpy as np
from astropy.modeling.models import Sersic2D
from astropy.convolution import convolve, convolve_fft
from photutils.aperture import aperture_photometry
from photutils.aperture import CircularAperture
from pysersic.priors import autoprior
import pysersic.loss as loss
from pysersic import SourceProperties
from pysersic import FitSingle
from jax.random import PRNGKey
import webbpsf
from pysersic.rendering import HybridRenderer
from pysersic.results import plot_residual
from matplotlib import pyplot as plt

### generate psf
fov = 5 ###arcsec
os_factor = 1


nc = webbpsf.NIRCam()
nc.options['parity'] = 'odd'

### This one is for FOV determined by angular size
psf_os = nc.calc_psf(fov_arcsec=fov, oversample=os_factor)[0].data

### This one is determined by number of pixels

## set up grid
npx_os = len(psf_os)
px_os_axis = np.arange(npx_os)
xpx_os, ypx_os = np.meshgrid(px_os_axis,px_os_axis)

px_scale_os = fov/npx_os

### define phot aperture for scaling
aper_rad_px = 0.5* 0.36/px_scale_os
aperture = CircularAperture((npx_os/2 - 1, npx_os/2 - 1), aper_rad_px)

### try just fitting psf+noise

raw_flux_pointsource = aperture_photometry(psf_os, aperture)['aperture_sum'][0]

pointsource_uJypx2 = (psf_os/raw_flux_pointsource)*3
pointsource_uJyacsc2 =  pointsource_uJypx2 / (px_scale_os**2)
pointsource_cps = pointsource_uJyacsc2/73.00475999916603
pointsource_counts = pointsource_cps * 6184.379999999998 ### this is the effective exptime for the mosaic I have open
    
read_noise = np.random.normal(scale=13.25, size=pointsource_counts.shape)  ### like JWST NIRCAM long channels? ### 0.75 is QE
poisson_noise = np.random.poisson(pointsource_counts) - pointsource_counts

pointsource_noisy = pointsource_counts + poisson_noise + read_noise


### calculate rms image
read_noise = np.random.normal(scale=13.65, size=pointsource_counts.shape)  ### like JWST NIRCAM long channels? ### 0.75 is QE
poisson_noise = np.random.poisson(pointsource_counts) - pointsource_counts
rms_im = np.sqrt((poisson_noise)**2+ read_noise**2)

prior  = autoprior(image = pointsource_noisy, profile_type = 'pointsource', sky_type = 'none')

sp = SourceProperties(pointsource_noisy)
sp.visualize()
prior.set_gaussian_prior('xc',sp.xc_guess,0.001) ### fixed
prior.set_gaussian_prior('yc',sp.yc_guess,0.001) ### fixed

fitter = FitSingle(data=pointsource_noisy,rms=rms_im,psf=psf_os,prior=prior,loss_func=loss.student_t_loss, 
                   renderer=HybridRenderer)


res = fitter.estimate_posterior(rkey = PRNGKey(1001), method='laplace')
sum_table = fitter.svi_results.summary()

dict = {}
for a, b in zip(sum_table.index, sum_table["mean"]):
    dict[a] = b

#bf_model = HybridRenderer(psf_os.shape, np.array(psf_os.astype(np.float32)),os_pixel_size=0,num_os=1).render_source(dict, profile_type="pointsource")
bf_model = HybridRenderer(psf_os.shape, np.array(psf_os.astype(np.float32))).render_source(dict, profile_type="pointsource")

fig, ax = plot_residual(pointsource_noisy/65,bf_model/65,vmin=-1.5,vmax=1.5)

fig2 = fitter.svi_results.corner()

print(max(np.ndarray.flatten(pointsource_noisy)))
print(max(np.ndarray.flatten(np.array(bf_model))))


plt.figure()
plt.imshow((pointsource_noisy-bf_model))
plt.colorbar()

```

**Comment by prappleizer** [comment]
Hi Kelly, do you mind attaching a file (or emailing me) the ultimate PSF array created from the webb tool, as I do not have it?

**Comment by ke27whal** [comment]
Hi- sorry I missed the notification on this. Here's the compressed fits file!
[psf_os1_fov5.fits.zip](https://github.com/user-attachments/files/17545154/psf_os1_fov5.fits.zip)

**Comment by prappleizer** [comment]
Hi @ke27whal . I've played around with the data you've sent a little bit. 

I believe your issue is that you are "fixing" the central positions to be within a very (very) tight margin (SD 0.001) based on the initial guess from the source-properties tool, but the initial guess is not guaranteed by any means to be within 0.001 of the true value. In my testing, the "guess" for `xc`/`yc` was around 80.01/80.0, but if the true intrinsic center is indeed at 80,80, then the code can't actually get there with a prior of SD=0.001 for a guess of 80.01. 

A better test would be setting the "fixed" prior on `xc`,`yc` to be the actual values you know them to be, or to simply allow them to be fit. For example, when I run your same fit but simply let `xc` and `yc` have an SD of 0.05, I then get `xc/yc` outputs of 79.996+/- 0.001, which is considerably closer to the ground truth, and produces residuals which are not asymmetric to well within a tenth of a percent.  Similarly, I can simply set the prior to be 80+/- 0.001, and then of course the output best fit is well constrained (79.999+/- 0.001).

In general, if all you're trying to do here is show that you can *fit* a point source with your PSF and get good residuals, then there's no reason to fix `xc` and `yc`; in fact this sort of defeats the purpose, especially given that in reality, your knowledge of the position of a point source in the image won't be very exact. I find here in your example that even with a large prior (say, sigma of tenth of a pixel), `pysersic` recovers reasonable constraints on `xc` and `yc`, and the residuals don't appear to be asymmetric. I suspect your asymmetry was coming just from the initial guess on xc being high, and the fit not having the range to reach the actual best value. Remember that a gaussian prior guides toward the center, so you'll get very few samples 5 sigma away, as in your case. As a note, a uniform prior would also potentially make more sense for you -- just span the pixel you know the center to be in, like 79.5 - 80.5. Either way. 


Here's the residual I get (comparing the normalized noisy image to the normalized `bf_model`, just changing the SD on the position priors to a very liberal 0.1). 

<img width="533" alt="Screenshot 2024-11-11 at 11 32 40 AM" src="https://github.com/user-attachments/assets/95c2e5cf-cad7-4223-8ab3-4f625d68c3fe">

**Comment by prappleizer** [comment]
Closing this for now but get in touch if there's further issues!
